File: ui/src/lib/languageOrgDiscovery.ts
/**
 * Language-Organization Discovery Service (UI)
 * Client-side service to discover available language-organization combinations
 */

import { mapLanguageToCatalogCode } from './languageMapping.js';

export interface LanguageOrgOption {
  language: string; // e.g., "en", "es-419"
  displayName?: string; // e.g., "English", "Spanish (Latin America)"
  organizations: string[]; // e.g., ["unfoldingWord", "Door43-Catalog"]
}

export interface LanguageOrgDiscoveryResult {
  options: LanguageOrgOption[];
  metadata: {
    responseTime: number;
    timestamp: string;
    totalLanguages: number;
    totalOrganizations: number;
  };
}

// Cache for discovery results
let cachedResult: LanguageOrgDiscoveryResult | null = null;
let cacheTimestamp: number = 0;
const CACHE_TTL = 6 * 60 * 60 * 1000; // 6 hours

/**
 * Discover all available language-organization combinations
 * Uses server endpoint if available, otherwise queries catalog directly
 */
export async function discoverLanguageOrgs(
  useCache: boolean = true,
): Promise<LanguageOrgDiscoveryResult> {
  // Check cache first
  if (useCache && cachedResult && Date.now() - cacheTimestamp < CACHE_TTL) {
    return cachedResult;
  }

  try {
    // Try server endpoint first (if it exists)
    try {
      const response = await fetch("/api/discover-language-orgs");
      if (response.ok) {
        const data = await response.json();
        cachedResult = data;
        cacheTimestamp = Date.now();
        return data;
      }
    } catch (error) {
      // Endpoint doesn't exist yet, fall through to direct catalog query
      console.debug("Discovery endpoint not available, using direct query");
    }

    // Fallback: Query catalog directly
    const catalogUrl =
      "https://git.door43.org/api/v1/catalog/search?stage=prod&metadataType=rc&limit=1000";

    const catalogResponse = await fetch(catalogUrl);
    if (!catalogResponse.ok) {
      throw new Error(
        `Failed to query catalog: ${catalogResponse.status}`,
      );
    }

    const catalogData = (await catalogResponse.json()) as {
      data?: Array<{
        name: string;
        title?: string;
        language?: string;
        owner?: string;
        lang?: string;
      }>;
    };

    if (!catalogData.data || catalogData.data.length === 0) {
      throw new Error("No resources found in catalog");
    }

    // Map to track language -> organizations
    const languageMap = new Map<
      string,
      { organizations: Set<string>; displayName?: string }
    >();

    // Extract unique language-organization combinations
    for (const resource of catalogData.data) {
      // Get language code - try multiple fields
      const langCode =
        resource.language ||
        resource.lang ||
        (resource.name.match(/^([a-z]{2,3}(?:-[A-Z][a-z]{3})?)/i)?.[1]?.toLowerCase());

      if (!langCode) continue;

      // Map to catalog code (e.g., es -> es-419)
      const catalogLangCode = mapLanguageToCatalogCode(langCode);

      // Get organization/owner
      const org = resource.owner;
      if (!org) continue;

      // Initialize language entry if not exists (use catalog code)
      if (!languageMap.has(catalogLangCode)) {
        // Try to extract display name from resource title
        let displayName: string | undefined;
        if (resource.title) {
          // Common patterns: "English ULT", "Spanish (Latin America) GLT"
          const titleMatch = resource.title.match(/^([^®(]+)/);
          if (titleMatch) {
            displayName = titleMatch[1].trim();
            // Remove resource type suffixes (ULT, UST, etc.)
            displayName = displayName.replace(
              /\s+(ULT|UST|GLT|GST|TN|TW|TQ|TA|TWL)$/i,
              "",
            );
          }
        }

        languageMap.set(catalogLangCode, {
          organizations: new Set(),
          displayName,
        });
      }

      // Add organization to language's set (use catalog code)
      languageMap.get(catalogLangCode)!.organizations.add(org);
    }

    // Convert to result format
    const options: LanguageOrgOption[] = Array.from(
      languageMap.entries(),
    ).map(([language, data]) => ({
      language,
      displayName: data.displayName,
      organizations: Array.from(data.organizations).sort(),
    }));

    // Sort by language code
    options.sort((a, b) => a.language.localeCompare(b.language));

    // Calculate totals
    const allOrgs = new Set<string>();
    options.forEach((opt) => {
      opt.organizations.forEach((org) => allOrgs.add(org));
    });

    const result: LanguageOrgDiscoveryResult = {
      options,
      metadata: {
        responseTime: 0, // Client-side doesn't track this
        timestamp: new Date().toISOString(),
        totalLanguages: options.length,
        totalOrganizations: allOrgs.size,
      },
    };

    // Cache result
    cachedResult = result;
    cacheTimestamp = Date.now();

    return result;
  } catch (error) {
    console.error("Failed to discover language-org combinations", error);
    throw error;
  }
}

/**
 * Get organizations for a specific language
 */
export function getOrganizationsForLanguage(
  discoveryResult: LanguageOrgDiscoveryResult,
  language: string,
): string[] {
  const option = discoveryResult.options.find(
    (opt) => opt.language === language,
  );
  return option?.organizations || [];
}

/**
 * Get display name for a language
 */
export function getLanguageDisplayName(
  discoveryResult: LanguageOrgDiscoveryResult,
  language: string,
): string {
  const option = discoveryResult.options.find(
    (opt) => opt.language === language,
  );
  return option?.displayName || language.toUpperCase();
}

File: examples/python-chatbot/language_org_discovery.py
"""
Language-Organization Discovery for Python Chatbot
Discovers available language-organization combinations from Door43 catalog
"""

import httpx
from typing import List, Dict, Optional

def map_language_to_catalog_code(language: str) -> str:
    """Map a language code to its catalog equivalent (e.g., es -> es-419)"""
    language_map = {
        'es': 'es-419',  # Spanish -> Latin American Spanish
        'es-MX': 'es-419',
        'es-AR': 'es-419',
        'es-CO': 'es-419',
        'es-CL': 'es-419',
        'es-PE': 'es-419',
    }
    return language_map.get(language, language)

async def discover_language_orgs() -> List[Dict[str, any]]:
    """
    Discover all available language-organization combinations from Door43 catalog
    
    Returns:
        List of dictionaries with 'language', 'displayName', and 'organizations' keys
    """
    catalog_url = "https://git.door43.org/api/v1/catalog/search?stage=prod&metadataType=rc&limit=1000"
    
    async with httpx.AsyncClient(timeout=30.0) as client:
        response = await client.get(catalog_url)
        response.raise_for_status()
        catalog_data = response.json()
    
    if not catalog_data.get("data"):
        return []
    
    # Map to track language -> organizations
    language_map = {}
    
    for resource in catalog_data["data"]:
        # Get language code - try multiple fields
        lang_code = (
            resource.get("language") or
            resource.get("lang") or
            (lambda: None if not resource.get("name") else 
             __import__("re").match(r"^([a-z]{2,3}(?:-[A-Z][a-z]{3})?)", resource["name"], __import__("re").IGNORECASE) and
             __import__("re").match(r"^([a-z]{2,3}(?:-[A-Z][a-z]{3})?)", resource["name"], __import__("re").IGNORECASE).group(1).lower()
            )()
        )
        
        if not lang_code:
            continue
        
        # Map to catalog code (e.g., es -> es-419)
        catalog_lang_code = map_language_to_catalog_code(lang_code)
        
        # Get organization/owner
        org = resource.get("owner")
        if not org:
            continue
        
        # Initialize language entry if not exists (use catalog code)
        if catalog_lang_code not in language_map:
            # Try to extract display name from resource title
            display_name = None
            if resource.get("title"):
                import re
                title_match = re.match(r"^([^®(]+)", resource["title"])
                if title_match:
                    display_name = title_match.group(1).strip()
                    # Remove resource type suffixes
                    display_name = re.sub(r"\s+(ULT|UST|GLT|GST|TN|TW|TQ|TA|TWL)$", "", display_name, flags=re.IGNORECASE)
            
            language_map[lang_code] = {
                "organizations": set(),
                "displayName": display_name
            }
        
        # Add organization to language's set (use catalog code)
        language_map[catalog_lang_code]["organizations"].add(org)
    
    # Convert to result format
    options = []
    for language, data in sorted(language_map.items()):
        options.append({
            "language": language,
            "displayName": data["displayName"],
            "organizations": sorted(list(data["organizations"]))
        })
    
    return options

def get_organizations_for_language(discovery_result: List[Dict], language: str) -> List[str]:
    """Get organizations for a specific language"""
    for option in discovery_result:
        if option["language"] == language:
            return option["organizations"]
    return []

def get_language_display_name(discovery_result: List[Dict], language: str) -> str:
    """Get display name for a language"""
    for option in discovery_result:
        if option["language"] == language:
            return option.get("displayName") or language.upper()
    return language.upper()

